src/v/archival/ntp_archiver_service.cc

```diff
@@ -508,6 +508,48 @@ ss::future<> ntp_archiver::sync_manifest_until_term_change() {
     }
 }
 
+ss::future<> ntp_archiver::unsafe_reset_metadata(ss::input_stream<char> is) {
+    // Deserialize outside of lock.
+    cloud_storage::partition_manifest req_m{};
+    co_await req_m.update(std::move(is));
+
+    // We're about to blow away the manifest. Lock it so the upload loop et al
+    // don't try uploading while this is happening.
```

Comment:
I'm not seeing the `_manifest_lock` being grabbed here. Is there something subtle that I'm missing?
Also, I'm not sure how the locking would work. You'd need to release the lock before the call to `co_await builder.replicate`, but other commands may be replicated before `reset_metadata_cmd` at that scheduling point.

Reply:
My bad, I removed an explicit acquisition of the manifest lock here because it gets taken during apply. The point of the lock is to make decisions/access to the manifest atomic with respect to one another, and since there's no real business logic in resetting the manifest (e.g. finding a segment based on the manifest, uploading, etc, as in the archiver loop), we needn't hold it longer than the duration of the apply of the records.

Reply:
Double oops, was conflating this with upload units acquisition. We could probably take that here to avoid races with the upload path.